File: [0685][Hard][Redundant_Connection_II]/Redundant_Connection_II.py
# -*- coding: utf-8 -*-
# leetcode time     cost : 88 ms
# leetcode memory   cost : 14.1 MB


class Solution:
    def findRedundantDirectedConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        father = list(range(n+1))
        indegree = [0] * (n+1)

        def find(x):
            f = x
            while f != father[f]:
                f = father[f]
            while f != x:
                x, father[x] = father[x], f
            return f

        def union(x, y):
            father[find(y)] = find(x)
        lasta, lastb, dupa, dupb = -1, -1, -1, -1
        for a, b in edges:
            indegree[b] += 1
            if indegree[b] == 2:
                dupa, dupb = a, b
            fa, fb = find(a), find(b)
            if fa != fb:
                union(a, b)
            else:
                lasta, lastb = a, b
        if dupb == -1:
            # 如果发现了环，还没发现入度为2的点，说明是一个顺序环，随便删
            return [lasta, lastb]
        # 发现了环，并且是一个有冲突的环 (环内存在不同方向的边)，得考虑删哪个边，先删除最后一条边试试
        father, firsta = list(range(n+1)), -1
        for a, b in edges:
            # 遍历除 2个入度为2的边 以外的所有边，
            if b != dupb:
                fa, fb = find(a), find(b)
                if fa != fb:
                    union(a, b)
                # 此时如果可以发现新加入的一条边对应的2个node有公共祖先，则说明此时存在 环 或 存在一个合法的树，
                # 则说明第一次遇到的2个入度的边[firsta, dupb]在环内，可以删除这条边，
                # 否则若没遇到公共祖先，则说明第2次遇到的边[dupa, dupb]在环内，可以删除这条边[dupa, dupb]
                else:
                    return [firsta, dupb]
            # 向father中加入第一次出现的2个入度的边，检测是否存在 树 或 环
            elif a != dupa:
                # 指向冲突节点的第一条边
                firsta = a
                fa, fb = find(a), find(b)
                if fa != fb:
                    union(a, b)
                # 此时如果可以发现新加入的一条边对应的2个node有公共祖先，则说明此时存在 环 或 存在一个合法的树，
                # 则说明第一次遇到的2个入度的边[firsta, dupb]在环内，可以删除这条边，
                # 否则若没遇到公共祖先，则说明第2次遇到的边[dupa, dupb]在环内，可以删除这条边[dupa, dupb]
                else:
                    return [firsta, dupb]
        # 删除 [dupa, dupb] 就没环了，这样就是结果
        return [dupa, dupb]


def main():
    graph1 = [[1, 2], [1, 3], [2, 3]]  # expect is [2,3],example 1,
    graph2 = [[1, 2], [2, 3], [3, 4], [4, 1],
              [1, 5]]  # expect is [4,1],example 2,
    obj = Solution()
    res = obj.findRedundantDirectedConnection(graph1)
    print("return value is ", res)


if __name__ == '__main__':
    main()

File: [0685][Hard][Redundant_Connection_II]/Redundant_Connection_II.java
// leetcode time     cost : 1 ms
// leetcode memory   cost : 39.9 MB
// https://leetcode.com/problems/redundant-connection-ii/discuss/278105/topic
class Solution {
    int[] anc;// 并查集
    int[] parent;// record the father of every node to find the one with 2
                 // fathers,记录每个点的父亲，为了找到双入度点

    public int[] findRedundantDirectedConnection(int[][] edges) {
        anc = new int[edges.length + 1];
        parent = new int[edges.length + 1];
        int[] edge1 = null;
        int[] edge2 = null;
        int[] lastEdgeCauseCircle = null;
        for (int[] pair : edges) {
            int u = pair[0];
            int v = pair[1];

            if (anc[u] == 0)
                anc[u] = u;
            if (anc[v] == 0)
                anc[v] = v;// init the union-find set 初始化并查集

            if (parent[v] != 0) {// node v already has a father, so we just skip the union of this edge and check
                                 // if there will be a circle ，跳过 edge2,并记下 edge1,edge2
                edge1 = new int[] { parent[v], v };
                edge2 = pair;
            } else {
                parent[v] = u;
                int ancU = find(u);
                int ancV = find(v);
                if (ancU != ancV) {
                    anc[ancV] = ancU;
                } else { // meet a circle , 碰到了环
                    lastEdgeCauseCircle = pair;
                }
            }
        }
        // 如果是情况2、3，则根据有没有碰到环返回 edge1 或 edge2
        if (edge1 != null && edge2 != null)
            // 第2次遇到的2个入度的边没有加入 parent[]，anc[]时，
            // 如果形成 环 或 树（有公共祖先），则说明第一次遇到的2入度边在环内，删除该边，
            // 否则第2次遇到的边在环内，删除edge2
            return lastEdgeCauseCircle == null ? edge2 : edge1;
        else
            return lastEdgeCauseCircle; // 否则就是情况1，返回那个导致环的最后出现的边。
    }

    private int find(int node) {
        if (anc[node] == node)
            return node;
        anc[node] = find(anc[node]);
        return anc[node];
    }
}